### PR TITLE
[TS][URGENT] LPS-158613

### PR DESCRIPTION
**Pull request by mbowerman**
https://issues.liferay.com/browse/LPS-158613

### CODE REVIEW

File: modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java [hunk 1]
```diff
@@ -71,26 +71,36 @@ public class S3FileCache {
 	public void cleanUpCacheFiles() {
 		_calledCleanUpCacheFilesCount++;
 
-		if (_calledCleanUpCacheFilesCount <
-				_cacheDirCleanUpFrequency.intValue()) {
+		if ((_calledCleanUpCacheFilesCount <
```
Comment (dantewang) [inline modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java:74]: @mbowerman This counter is thread-unsafe, as it's a plain int. I would recommend to change it to `AtomicInteger` or `Semaphor`.

File: modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java [hunk 2]
```diff
@@ -71,26 +71,36 @@ public class S3FileCache {
...
+				_cacheDirCleanUpFrequency.intValue()) ||
+			_cacheCleanUpRunning) {
 
 			return;
 		}
 
 		synchronized (this) {
```
Comment (dantewang) [inline modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java:81]: @mbowerman I think the problem here is about the usage of `synchronized (this)`, as this is a pure waiting. I think we can avoid this problem by using `Lock.tryLock()` as it's "check-and-skip":

```
if (_lock.tryLock()) {
    try {
        ...
    }
    finally {
        _lock.unlock();
    }
}

...

private final Lock _lock = new ReentrantLock();
```

File: modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java [hunk 3]
```diff
@@ -71,26 +71,36 @@ public class S3FileCache {
...
+				_cacheCleanUpRunning = true;
+
+				if (_calledCleanUpCacheFilesCount <
+						_cacheDirCleanUpFrequency.intValue()) {
+
+					return;
+				}
 
-			_calledCleanUpCacheFilesCount = 0;
+				_calledCleanUpCacheFilesCount = 0;
```
Comment (dantewang) [inline modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java:91]: @mbowerman When using `tryLock`, I think we can remove the if-statement here, and reset the counter after clean-up finishes.

**Review by dantewang — CHANGES_REQUESTED** [review]
Hi @mbowerman, please see my comment and make changes. Also, could you add some tests to reveal/demonstrate the issue. Feel free to let me know if you have any concern. Thanks!

File: modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java (hunk 2 above)
Comment (mbowerman) [reply modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java]: Ah, perfect! I forgot about ReentrantLock. I like this solution better. Done.

File: modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java (hunk 3 above)
Comment (mbowerman) [reply modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java]: Done. Thanks for the quick review!

**Comment by dantewang** [comment]
@mbowerman Thanks for the update! Please make the counter an `AtomicInteger` and try to add tests.

**Comment by mbowerman** [comment]
Hi @dantewang,

Sorry, I wasn't able to get to this today. I'll try to get to it tomorrow.

**Comment by mbowerman** [comment]
Unfortunately it does not look like I will be able to get to this today either. I hope to be able to get to it on Monday.

**Comment by mbowerman** [comment]
Hi @dantewang,

Sorry for the delay on this. I have made the requested changes and pushed them up. The unit test is a bit ugly because I wasn't sure of the best way to test a multithreaded bug like this. Let me know if you have any questions or feedback about it.

**Comment by dantewang** [comment]
@mbowerman Thanks! I discussed this issue with Tina today, and we will look into the possibility for a different solution -- for example, using scheduled/background task. So this PR will be held for a bit longer. Will let you know when we get any update.

**Comment by tinatian** [comment]
@dantewang I am going to close this pull, please resend.

Thanks.
Tina

**Pull request closed without merging**